rpm-ostree kargs command

peterbaouoft opened the pull request:
This is initial groundwork for https://github.com/projectatomic/rpm-ostree/issues/594.

This commit sets up most of the required
front end logic( arg parsing, transaction handling), and will
be used in the following commits.

There is nothing really fancy in this commit, as most of the code
shares the similar style between other dbus related commands.

cgwalters commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -0,0 +1,135 @@
+
+  g_variant_dict_init (&dict, NULL);
+  g_variant_dict_insert (&dict, "merge", "b", opt_merge);
+  g_variant_dict_insert (&dict, "editor", "b", opt_editor);
```
Editor should be purely client side, right?  I can't think of a reason to send it to the daemon offhand.

cgwalters commented on src/daemon/org.projectatomic.rpmostree1.xml:
```diff
@@ -278,6 +278,18 @@
+        "editor" (type 'b')
+        "merge" (type 'b')
+    -->
+    <method name="KernelArgs">
```
This feels close to the `SetInitramfsState` method - I don't see a reason for it to be experimental either.  So let's move this right underneath that method in the XML, and also implement it close by in `rpmostreed-os.c`?

We might actually consider a general `SetBootConfiguration` method, but we don't need to do that right now.  Maybe though internally route `os_handle_set_initramfs_state()` and `os_handle_set_boot_args()` into an internal-only `os_handle_boot_configuration()` method or so at least?  Like how we route the pkg install/deploy etc. into a generic `os_merge_or_start_deployment_txn()`.

(And really perhaps down the line we need to be able to both do initramfs/kargs/deploy all in one, but no need to worry about that now)

cgwalters commented on src/daemon/org.projectatomic.rpmostree1.xml:
```diff
@@ -278,6 +278,18 @@
+    <method name="KernelArgs">
+      <arg type="as" name="kernel_args_added" direction="in"/>
+      <arg type="as" name="kernel_args_replaced" direction="in"/>
+      <arg type="as" name="kernel_args_removed" direction="in"/>
```
The semantics here get interesting because kernel arguments are [a multimap](https://en.wikipedia.org/wiki/Multimap).   For example dracut parses `rd.lvm.lv=foo/bar rd.lvm.lv=baz/moo` etc.  So if I specify `rd.lvm.lv=hello/world` in `kernel_args_replaced`, does that eliminate all of the other `rd.lvm.lv` args?

cgwalters replied:
There's also a similar discussion for ostree remotes in https://github.com/ostreedev/ostree/pull/1166#issuecomment-332010464

peterbaouoft replied:
>So if I specify rd.lvm.lv=hello/world in kernel_args_replaced, does that eliminate all of the other rd.lvm.lv args

I believe so, because my original plan for backend logic was to reuse the parsing functions for OstreeKernelArgs( `_ostree_kernel_args_append_kargv` and `_ostree_kernel_args_replace_kargv` ) to parse all of the kernel arguments. And the old kernel arg parsing logic uses a  hashtable to replace the duplicate keys, so other rd.lvm.lv args existed earlier will be removed. 

This is the original code https://github.com/ostreedev/ostree/blob/master/src/ostree/ot-admin-instutil-builtin-set-kargs.c#L97 and https://github.com/ostreedev/ostree/blob/master/src/libostree/ostree-kernel-args.c#L100

Do we want that?

peterbaouoft replied:
**tl;dr**: Realized `_ostree_kernel_args_append_kargv()` might not be usable

After a bit more look into the code base, I realized some of the code in ostree code base does not function exactly as I thought T_T. Sorry for that and  kinda take back some of what I said earlier. Below is their logic for appending_kargs().
https://github.com/ostreedev/ostree/blob/master/src/libostree/ostree-kernel-args.c#L131

>So if I specify rd.lvm.lv=hello/world in kernel_args_replaced, does that eliminate all of the other rd.lvm.lv args

If we were to adopt the `_ostree_kernel_args_append_kargv()` , `rd.lvm.lv=hello/world` will instead be the key, and `hello/world` will be the value. So, before.. if we do a replacement by calling `_ostree_kernel_args_replace_take() `. Other `rd.lvm.lv` values will not be affected because they do not have the exact string...............

So, I am thinking I possibly need to reformat the `_ostree_kernel_args_append` function because.. having `key=value` string as a key for the KARGS hashtable seems not right to me. But, I still think replace function in ostree end should be usable. Thoughts ? And thanks for the review comments. :)

peterbaouoft commented on src/libpriv/rpmostree-kargs-process.c:
```diff
@@ -0,0 +1,331 @@
+/*
```
Note: This file is copied straight from https://github.com/ostreedev/ostree/blob/master/src/libostree/ostree-kernel-args.c, just for testing the functionality purposes. Need to decide whether or not we should make the ostree-kernel-args public.

peterbaouoft replied:
Also Note: Functions I newly added will be commented with `Note: this function is newly added to the API` on top of the function. The other functions without those are from the old existing api.

peterbaouoft commented on src/libpriv/rpmostree-kargs-process.h:
```diff
@@ -0,0 +1,60 @@
+/*
```
Note: Also same, copied from ostree for testing purposes

peterbaouoft commented:
Hey, there are a few design decisions that I am currently uncertain about, and like to ask them here. So.. thanks in advance :p. 

1: Do we want to make the functions/API in ostree kargs public, or it is fine by copying these to rpm-ostree side directly. Reason why I asked is I am also adding some functions to the existing API to achieve 'delete' and 'replace' related functionality, and unsure where to put it.

2: For the replace option. There is a special case that there could be multiple values associated with one key. In that case, we only want the user to replace one of them, and we may want the user to type replace the key by providing `key=value=new_value`. But it feels a bit messy that way because of the two equal signs.. so wondering is it fine that I go with (`rpm-ostree ex kargs --replace key=value=new_value`) implementation for now? 

3: When we do append/replace option, there could be a case that users append or replace a kernel argument with exact key/value pair. e.g: `rpm-ostree ex kargs --append hi=hi --append hi=hi`. Do we need/want to worry about that case? 

@cgwalters  thoughts?

cgwalters commented:
For question 1) I'm fine copying them into rpm-ostree for now.   We can always dedup later.  But it's up to you...we can add APIs to libostree too.  Maybe on the `OstreeDeployment` class?

For question 2) `--replace key=value=new_value` seems good to me!

For question 3) Let's not make that an error...though it is yet another corner case.  Hmm, I wonder if anything in the kernel actually uses that.  Probably not...but I'm not 100% sure.

peterbaouoft commented:
Currently the main functionality should be done (I think) and here is the introduction to some of the functionalities. I am sure there are still many things that are not polished, and the comments/commit messages need to be reworked too, but wanted to see what your opinions are first :). I am still an inexperienced coder, so there might still be a lot things that I did wrong, so feel free to point out the flaws/errors/bad code out =). And most importantly, thanks for your time :). 

Here is the introduction to the functionalities :D: Note, by default, the kernel arguments are first collected/merged from the first/pending deployment' kernel args in the system.  
==============================Kargs=====================================
```
[rbao@unused-10-15-17-147 f26AH]$ vagrant ssh
Last login: Thu Oct  5 18:27:15 2017 from 192.168.121.1
[vagrant@localhost ~]$ sudo rpm-ostree status
State: idle
Deployments:
* vmcheck
                 Timestamp: 2017-10-05 18:38:22
                    Commit: f034e215cce4fa667cdfb53692232da947577e81d61ba4ac892cbd133aaaf980
[vagrant@localhost ~]$ sudo rpm-ostree ex kargs
The kernel arguments are:
no_timer_check console=tty1 console=ttyS0,115200n8 net.ifnames=0 biosdevname=0 rd.lvm.lv=atomicos/root root=/dev/mapper/atomicos-root ostree=/ostree/boot.1/fedora-atomic/aac51b16bc712956159c198c65180ef483f9dc8e0dd74f672f91bb365da5822b/0 hi=hiii

[vagrant@localhost ~]$ grep ^options /boot/loader/entries/ostree-fedora-atomic-0.conf  
options no_timer_check console=tty1 console=ttyS0,115200n8 net.ifnames=0 biosdevname=0 rd.lvm.lv=atomicos/root root=/dev/mapper/atomicos-root ostree=/ostree/boot.1/fedora-atomic/aac51b16bc712956159c198c65180ef483f9dc8e0dd74f672f91bb365da5822b/0 hi=hiii

```

===============================Kargs append===============================
```
[vagrant@localhost ~]$ grep ^options /boot/loader/entries/ostree-fedora-atomic-0.conf  
options no_timer_check console=tty1 console=ttyS0,115200n8 net.ifnames=0 biosdevname=0 rd.lvm.lv=atomicos/root root=/dev/mapper/atomicos-root ostree=/ostree/boot.1/fedora-atomic/aac51b16bc712956159c198c65180ef483f9dc8e0dd74f672f91bb365da5822b/0 hi=hiii
[vagrant@localhost ~]$ sudo rpm-ostree status
State: idle
Deployments:
* vmcheck
                 Timestamp: 2017-10-05 18:38:22
                    Commit: f034e215cce4fa667cdfb53692232da947577e81d61ba4ac892cbd133aaaf980
// append will create a new pending deployment
[vagrant@localhost ~]$ sudo rpm-ostree ex kargs --append test=test
Copying /etc changes: 24 modified, 0 removed, 141 added
Transaction complete; bootconfig swap: yes deployment count change: 1
[vagrant@localhost ~]$ grep ^options /boot/loader/entries/ostree-fedora-atomic-0.conf  
options no_timer_check console=tty1 console=ttyS0,115200n8 net.ifnames=0 biosdevname=0 rd.lvm.lv=atomicos/root root=/dev/mapper/atomicos-root ostree=/ostree/boot.0/fedora-atomic/aac51b16bc712956159c198c65180ef483f9dc8e0dd74f672f91bb365da5822b/0 hi=hiii test=test
[vagrant@localhost ~]$ sudo rpm-ostree status
State: idle
Deployments:
  vmcheck
                 Timestamp: 2017-10-05 18:38:22
                    Commit: f034e215cce4fa667cdfb53692232da947577e81d61ba4ac892cbd133aaaf980

* vmcheck
                 Timestamp: 2017-10-05 18:38:22
                    Commit: f034e215cce4fa667cdfb53692232da947577e81d61ba4ac892cbd133aaaf980

// test multiple appends
[vagrant@localhost ~]$ sudo rpm-ostree ex kargs --append hello --append test=again --append=this=test
Copying /etc changes: 24 modified, 0 removed, 141 added
Transaction complete; bootconfig swap: yes deployment count change: 0
[vagrant@localhost ~]$ grep ^options /boot/loader/entries/ostree-fedora-atomic-0.conf
options no_timer_check console=tty1 console=ttyS0,115200n8 net.ifnames=0 biosdevname=0 rd.lvm.lv=atomicos/root root=/dev/mapper/atomicos-root ostree=/ostree/boot.1/fedora-atomic/aac51b16bc712956159c198c65180ef483f9dc8e0dd74f672f91bb365da5822b/0 hi=hiii test=test test=again hello this=test

// Finally we can do a rollback
[vagrant@localhost ~]$ sudo rpm-ostree rollback 
Moving 'f034e215cce4fa667cdfb53692232da947577e81d61ba4ac892cbd133aaaf980.0' to be first deployment
Transaction complete; bootconfig swap: yes deployment count change: 0
Run "systemctl reboot" to start a reboot
[vagrant@localhost ~]$ grep ^options /boot/loader/entries/ostree-fedora-atomic-0.conf
options no_timer_check console=tty1 console=ttyS0,115200n8 net.ifnames=0 biosdevname=0 rd.lvm.lv=atomicos/root root=/dev/mapper/atomicos-root ostree=/ostree/boot.0/fedora-atomic/aac51b16bc712956159c198c65180ef483f9dc8e0dd74f672f91bb365da5822b/0 hi=hiii
```

===========================Kargs Delete===================================
```
[vagrant@localhost ~]$ grep ^options /boot/loader/entries/ostree-fedora-atomic-0.conf
options no_timer_check console=tty1 console=ttyS0,115200n8 net.ifnames=0 biosdevname=0 rd.lvm.lv=atomicos/root root=/dev/mapper/atomicos-root ostree=/ostree/boot.0/fedora-atomic/aac51b16bc712956159c198c65180ef483f9dc8e0dd74f672f91bb365da5822b/0 hi=hiii
// delete one to one key/value pair
[vagrant@localhost ~]$ sudo rpm-ostree ex kargs --delete hi --delete no_timer_check
Copying /etc changes: 24 modified, 0 removed, 142 added
Transaction complete; bootconfig swap: yes deployment count change: 0
[vagrant@localhost ~]$ grep ^options /boot/loader/entries/ostree-fedora-atomic-0.conf
options console=tty1 console=ttyS0,115200n8 net.ifnames=0 biosdevname=0 rd.lvm.lv=atomicos/root root=/dev/mapper/atomicos-root ostree=/ostree/boot.1/fedora-atomic/aac51b16bc712956159c198c65180ef483f9dc8e0dd74f672f91bb365da5822b/0

//delete key with multiple values: 
[vagrant@localhost ~]$ sudo rpm-ostree ex kargs --delete console
error: Unable to delete console with multiple values associated with it, see rpm-ostree ex kargs -h for usage
[vagrant@localhost ~]$ sudo rpm-ostree ex kargs --delete console=tty1
Copying /etc changes: 24 modified, 0 removed, 142 added
Transaction complete; bootconfig swap: yes deployment count change: 0
[vagrant@localhost ~]$ grep ^options /boot/loader/entries/ostree-fedora-atomic-0.conf
options console=ttyS0,115200n8 net.ifnames=0 biosdevname=0 rd.lvm.lv=atomicos/root root=/dev/mapper/atomicos-root ostree=/ostree/boot.0/fedora-atomic/aac51b16bc712956159c198c65180ef483f9dc8e0dd74f672f91bb365da5822b/0
```
==============================Kargs Replace===================================
```
[vagrant@localhost ~]$ grep ^options /boot/loader/entries/ostree-fedora-atomic-0.conf
options no_timer_check console=tty1 console=ttyS0,115200n8 net.ifnames=0 biosdevname=0 rd.lvm.lv=atomicos/root root=/dev/mapper/atomicos-root ostree=/ostree/boot.1/fedora-atomic/aac51b16bc712956159c198c65180ef483f9dc8e0dd74f672f91bb365da5822b/0 hi=hiii
// single pair replacement
[vagrant@localhost ~]$ sudo rpm-ostree ex kargs --replace no_timer_check=hello --replace hi=hmm
Copying /etc changes: 24 modified, 0 removed, 142 added
Transaction complete; bootconfig swap: yes deployment count change: 0
[vagrant@localhost ~]$ grep ^options /boot/loader/entries/ostree-fedora-atomic-0.conf
options no_timer_check=hello console=tty1 console=ttyS0,115200n8 net.ifnames=0 biosdevname=0 rd.lvm.lv=atomicos/root root=/dev/mapper/atomicos-root ostree=/ostree/boot.1/fedora-atomic/aac51b16bc712956159c198c65180ef483f9dc8e0dd74f672f91bb365da5822b/0 hi=hmm

// try replace a key with multiple values
[vagrant@localhost ~]$ sudo rpm-ostree ex kargs --replace console    
error: Unable to replace console with multiple values associated with it, see rpm-ostree ex kargs -h for usage
[vagrant@localhost ~]$ sudo rpm-ostree ex kargs --replace console=tty1=test
Copying /etc changes: 24 modified, 0 removed, 142 added
Transaction complete; bootconfig swap: yes deployment count change: 0
[vagrant@localhost ~]$ grep ^options /boot/loader/entries/ostree-fedora-atomic-0.conf
options no_timer_check=hello console=test console=ttyS0,115200n8 net.ifnames=0 biosdevname=0 rd.lvm.lv=atomicos/root root=/dev/mapper/atomicos-root ostree=/ostree/boot.0/fedora-atomic/aac51b16bc712956159c198c65180ef483f9dc8e0dd74f672f91bb365da5822b/0 hi=hmm
```

peterbaouoft commented:
Rebased ⬆️ and fix most of the style issues that I spotted.  Lifting WIP. :)

The editor command still is left undone because I am hesitant whether I should do a copy paste too :(. It seems like the logic will be exact same as: https://github.com/ostreedev/ostree/blob/master/src/ostree/ot-editor.c#L59. But... I tried, and.. this function is builtin if I am doing it correctly..

In addition to the code duplicate, I am also unsure about which `RpmOstreeBuiltinFlags` I should actually give to the rpm-ostree ex kargs. As we also want to enforce the editor option to be under root privilege too.

Also, there are many... fixup commits. I decided not to squash them yet, in case you want to see the progress. If it is way too messy, let me know, and the patch should be ok for review :)

cgwalters commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -0,0 +1,143 @@
+  /* We do not want editor option to be handled via daemon */
+  if (opt_editor)
+    {
+      ;
```
Let's comment out the option until it's implemented?

cgwalters commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -0,0 +1,143 @@
+      char *empty_strv[] = {NULL};
+
+      /* dbus does not allow empty string arrays, assign it
+       * to be empty string array here to prevent erroring out */
```
"does not allow NULL to mean the empty string array" would be a bit more precise.

cgwalters commented on src/daemon/rpmostree-sysroot-upgrader.c:
```diff
@@ -1030,6 +1032,31 @@ rpmostree_sysroot_upgrader_prep_layering (RpmOstreeSysrootUpgrader *self,
+ *
+ */
+gboolean
+rpmostree_sysroot_upgrader_deploy_set_kargs (RpmOstreeSysrootUpgrader *self,
```
I'm OK with this, but I think down the line we should add the kernel args to `rpmostree_sysroot_upgrader_deploy()` and have the other callers pass `NULL`.

peterbaouoft replied:
hey thanks for the review :). For this part, I believe other callers are still able to set NULL for the kernel args, as the kargs_strv is implemented as one of the attributes from the upgrader, similar to other attributes as well. 
https://github.com/projectatomic/rpm-ostree/pull/1013/files#diff-af8b24e3bf31175f007dd70d4b1a46fcR83

My personal opinion, but I could be wrong, is .. it feels a bit unnecessary to change other code when you can just set NULL to the kargs_strv attribute prior to calling `rpmostree_sysroot_upgrader_deploy`? ( I kinda tried to avoid changing too much existing code base too =) ). But I am also fine if you want to change the function.

cgwalters commented on src/daemon/rpmostreed-transaction-types.c:
```diff
@@ -1389,3 +1390,169 @@ rpmostreed_transaction_new_refresh_md (GDBusMethodInvocation *invocation,
+  /* Set this boolean to have a default value */
+  gboolean import_from_cmd = FALSE;
+
+  if (self->flags & RPMOSTREE_TRANSACTION_KERNEL_ARG_FLAG_IMPORT_CMD)
```
I read "import cmd" as an abbreviation for "import command" - let's say "_IMPORT_CMDLINE" to be clearer that it's the "command line"?  Or maybe to be even more clear, say "FLAG_IMPORT_PROC_CMDLINE" ?  And name the boolean `import_from_proc_cmdline`.

cgwalters commented:
While it's sometimes useful to see the development stream, at this point I think we should squash the patches together and do further fixups on top.

peterbaouoft commented on tests/vmcheck/test-kernel-args.sh:
```diff
@@ -0,0 +1,98 @@
+vm_cmd grep ^options /boot/loader/entries/ostree-$osname-0.conf > tmp_conf.txt
+assert_not_file_has_content 'REPLACE_MULTI_TEST=NUMBERTWO'
+assert_not_file_has_content 'REPLACE_MULTI_TEST=NEWTEST'
+assert_not_file_has_content 'APPENDARG=2NDAPPEND'
```
yup, thanks for catching it :). The weird part is it did not error out too.... Will be more careful next time.

peterbaouoft commented:
Added a fixup to address the comments above, ⬆️, the commits are also auto squashed. Do note that there are some extra code introduced to fix/improve the existing functionality.

1: in the src/libpriv/rpmostree-kargs-process.c, an extra corner case handlation was added to fix the replace logic

2: 
> I was thinking we add the kernel args to the DBus properties but then I realized that they might contain secret data. So perhaps we add a GetDeploymentBootconfig DBus method?

This was implemented in the fix up as well. (not sure if it is a good place to put it yet), but decided to put the entire implementation into the fix up as well because some of the duplicate code used in the earlier commits can be removed. Now, we will first load the existing kernel arguments outside the kernel arguments transaction and pass it into the transaction, instead of loading it directly inside the transaction. 

3: After the GetDeploymentBootconfig is implemented, a few extra options( reboot, deployid) were added in src/app/rpmostree-builtin-kargs.c. The import-from-cmd option is also no longer handled inside the transaction execute function. 

4: There might be many things I did wrong  that I am not aware of :(. So feel free to give comments so that I can fix them :). Thank you for your time :D!

cgwalters commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -0,0 +1,203 @@
+
+static GOptionEntry option_entries[] = {
+  { "os", 0, 0, G_OPTION_ARG_STRING, &opt_osname, "Operation on provided OSNAME", "OSNAME" },
+  { "deployid", 0, 0, G_OPTION_ARG_STRING, &opt_deployid, "Modify the kernel args from a specific deployment based on id. Id is in the form of 'osname-checksum.deployserialnum' ", "DEPLOYID"},
```
I see this is used in the DBus API too.  Hmm.  I'm not sure.  I think my tentative vote here is to use an index in the API too but I could definitely be convinced to try to standardize this "deployid" too.

peterbaouoft replied:
> I think my tentative vote here is to use an index in the API too

Sure, index here makes more sense. I used the deploy id implementation just simply because this was also used in the old DBus methods :p. Will do a change.

cgwalters commented on src/daemon/rpmostreed-os.c:
```diff
@@ -1128,6 +1130,154 @@ os_handle_set_initramfs_state (RPMOSTreeOS *interface,
+    }
+  OstreeBootconfigParser *bootconfig = ostree_deployment_get_bootconfig (target_deployment);
+
+  /* Note because boot config is a private structure.. currently I have no good way
```
This we should fix in libostree I think.  But it doesn't need to be done to land this change.

cgwalters commented on src/daemon/org.projectatomic.rpmostree1.xml:
```diff
@@ -220,6 +220,24 @@
+        "reboot" (type 'b')
+    -->
+    <method name="KernelArgs">
+      <arg type="s" name="existing_kernel_arg_string"/>
```
Is this value actually used by the daemon?  It looks to me like we pass it all the way down but don't actually use it, rather it's just parsed on the client side.   Is that correct?  If so, let's just drop this from the DBus API?

peterbaouoft replied:
Hey, thanks for the review :). We actually do use it, it is this line
``` 
__attribute__((cleanup(_ostree_kernel_args_cleanup))) OstreeKernelArgs *kargs = _ostree_kernel_args_from_string (self->existing_kernel_args);
```

We use this attribute so that we do not need to do the extra loading of the wanted kargs again in the transaction execute, as we have already done the loading in `rpmostree_os_call_get_deployment_boot_config_sync` function.

cgwalters replied:
Ah, right OK.

peterbaouoft commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -0,0 +1,203 @@
+
+  const char* deploy_str = opt_deployid ?: "";
+  g_autoptr(GVariant) boot_config = NULL;
+  if (!rpmostree_os_call_get_deployment_boot_config_sync (os_proxy,
```
Just noticed via a bit more testing, the policy kit verification will be called twice since there are two different os methods being called. Will do a bit more further investigation on how to avoid that from happening.

peterbaouoft commented:
Added a fixup to address the comments above, ⬆️, also added a fix for solving the duplicated policy kit authentication calls. 

But the implementation for the id parsing, and the solution for duplicate policy kit seems a bit ugly.. But, can't seem to think for a better solution atm. So, suggestions are welcome, and thanks for your time :)

cgwalters commented on src/app/rpmostree-builtin-kargs.c:
```diff
@@ -52,12 +52,13 @@ static GOptionEntry option_entries[] = {
 
   g_variant_dict_init (&dict, NULL);
   g_variant_dict_insert (&dict, "reboot", "b", opt_reboot);
+  g_variant_dict_insert (&dict, "authorized", "b", is_authorized_prior);
```
We can't do that - that entirely subverts the authorization check, as any caller could set that to `TRUE`.  What we should be using here is polkit's "imply" bits.  See [the docs](https://www.freedesktop.org/software/polkit/docs/latest/polkit.8.html).  And for example (on my workstatation under `/usr/share/polkit-1/actions`):

```
  <action id="org.fedoraproject.FirewallD1.all">
    <description>Firewall</description>
    <message>System policy prevents inspecting and changing firewall</message>
    <defaults>
      <allow_any>auth_admin_keep</allow_any>
      <allow_inactive>auth_admin_keep</allow_inactive>
      <allow_active>auth_admin_keep</allow_active>
    </defaults>
    <annotate key="org.freedesktop.policykit.imply">org.fedoraproject.FirewallD1.info org.fedoraproject.FirewallD1.config org.fedoraproject.FirewallD1.config.info org.fedoraproject.FirewallD1.direct org.fedoraproject.FirewallD1.direct.info org.fedoraproject.FirewallD1.policies org.fedoraproject.FirewallD1.policies.info</annotate>
  </action>

        <action id="org.freedesktop.login1.inhibit-block-shutdown">
                <description>Allow applications to inhibit system shutdown</description>
...
                <annotate key="org.freedesktop.policykit.imply">org.freedesktop.login1.inhibit-delay-sleep</annotate>
```

etc.

peterbaouoft replied:
Ah ha, that is what I was looking for :), thanks for the info, will do the change when I am back online.

peterbaouoft commented:
Pushed a fixup ⬆️ and added some tests. 

Tho there is still one thing I am slightly uncertain:

The current implementation for policy kit is that if the user is authenticated to read the boot config via `GetDeploymentBootConfig`, they will be authenticated to change boot configuration too. That includes "set initramfs" and "kernel args". It makes sense to authenticate for changing kernel args, but I am not sure we should give user permission to set initramfs too...

However, I am also a bit not sure if it is worth to separate those two actions apart yet, as "initramfs" and "kernel args" both involve changing boot configurations. So,...  Suggestions? and thanks for your time :).

cgwalters commented:
Hum...now that I look, at least on ostree-managed systems we write /boot/loader/entries files world-readable, and for that matter `/proc/cmdline` is world-readable too.   

Ah...I think it's a special GRUB thing, because grub supports e.g. `GRUB2_PASSWORD`.  So...we could simplify all of this and have the boot configuration be a no-authz-required method?  (Sorry about going back and forth here...we could also just keep it the way it is out of conservatism)

peterbaouoft commented:
> So...we could simplify all of this and have the boot configuration be a no-authz-required method?

yup, that makes more sense for ostree-managed systems. I assume the implementation would just be moving the existing `GetDeploymentBootConfig` to another place(be a normal function) so that it does not involve dbus/polkit? 

>Sorry about going back and forth here

ah no worries. I think the more comments we have for a patch, the fewer regression we will need in the future. And this definitely makes me learn as well  :) ( I thought ostree-managed systems will need admin priviledges for accessing boot loader related entries too).

cgwalters commented:
Well, we could keep the method as is I think, just drop the authorization check?  Or alternatively make it a property.

cgwalters commented:
bot, retest this please

cgwalters commented on src/libpriv/rpmostree-kargs-process.c:
```diff
@@ -0,0 +1,624 @@
+        }
+      /* Handle both no value case, and the case when inputting
+         key=value for a replacement */
+      else if ((!*val || (is_replaced && !*replaced_val)) &&
```
Maybe let's do e.g.
`const gboolean key_only = !*val` above?

peterbaouoft replied:
Sure.

cgwalters commented on src/libpriv/rpmostree-kargs-process.c:
```diff
@@ -0,0 +1,624 @@
+       * there is only one single key, and the second val
+       * will now represent the new value (no second split
+       * will happen this case) */
+      else if (val && is_replaced && values->len == 1)
```
`const gboolean single_value = values->len == 1` above too?

Hm...and does this need to be `if (*val && ...`?  (With the extra `*`?)  I

peterbaouoft replied:
> Hm...and does this need to be if (*val && ...? (With the extra *?)

yup, thanks for catching it. Did not notice it when I checked it.

cgwalters commented:
Some new comments; I think we're closing in here.  I'd feel better if we had some unit-style tests for the kernel argument parsing.  Should be pretty easy to do right?

peterbaouoft commented:
Yup, many thanks for the review :). 
> I'd feel better if we had some unit-style tests for the kernel argument parsing. Should be pretty easy to do right

Sure, I am going to start writing those, hopefully won't take so long. Haven't written a long enough unit-style tests yet :p.

One more thing, the editor functionality is also almost done, so it will be a bit weird to add fixup to the old commits along with it. (considering there are many fixups already). So do you mind I autosquash the fixup commits earlier( including the newly added editor commit). Then every fixup commits later will be built on top? @cgwalters

cgwalters commented:
Yeah, squashing now is totally fine!

cgwalters commented:
Or if you want we can do the editor as a follow-up?

peterbaouoft commented:
Thanks for the fast response! :). 
> Or if you want we can do the editor as a follow-up?

Sure, we can do it as a follow-up. I will first start writing the fixup commit and the unit test in that case :p.

peterbaouoft commented on src/libpriv/rpmostree-kargs-process.c:
```diff
@@ -0,0 +1,624 @@
+  if (!values)
+    {
+      g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                           "The key is not found, type rpm-ostree ex kargs to see what args are there");
```
Sure :)

peterbaouoft commented on src/libpriv/rpmostree-kargs-process.c:
```diff
@@ -0,0 +1,624 @@
+                            const char       *arg,
+                            GError           **error)
+{
+  /* Make a duplicate of the passed in arg for manipulation */
```
yup, nice catch!

peterbaouoft commented on src/libpriv/rpmostree-kargs-process.c:
```diff
@@ -0,0 +1,624 @@
+  if (!g_hash_table_remove (kargs->table, key))
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                   "Failed to remove %s from hashtable",
```
Sure .

peterbaouoft commented on src/libpriv/rpmostree-kargs-process.c:
```diff
@@ -0,0 +1,624 @@
+  if (!_ostree_ptr_array_find (kargs->order, key,
+                              &key_index))
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
```
yup, thanks for pointing it out.

peterbaouoft commented:
bot, retest this please

peterbaouoft commented:
Pushed a fixup ⬆️, addressed the comments mentioned above.  Also pushed a commit for unit tests. There a few points that I think are worth noting:

1: `rpmostreed_deployment_get_for_index` did not have a unit test with it. The ostree sysroot and deployment is hard to set up for the testing. Thus choosing not to do so.

2: Not sure if I need to do unit testing for the dbus/command line related functions. 

3: I also need to expose the kargs' table through `_ostree_kernel_arg_get_kargs_table` and `_ostree_kernel_arg_get_key_array`. So that might not be too neat, so suggestion for alternative implementation are always welcome :).

cgwalters commented:
OK, this all looks pretty good to me; there's some relatively minor stuff but not worth more round trips here.  Let's get it in - since this is under `ex` we can always do fixes as a followup.  Thanks a lot for working on this!

@rh-atomic-bot r+ aaf2ed1

peterbaouoft commented:
Many thanks for all of the meaningful review in the past month! :)
